core/src/main/java/org/apache/iceberg/util/ParallelIterable.java

```diff
@@ -257,17 +257,17 @@ private static class Task<T> implements Supplier<Optional<Task<T>>>, Closeable {
     @Override
     public Optional<Task<T>> get() {
       try {
+        if (queue.size() >= approximateMaxQueueSize) {
+          // Yield when queue is over the size limit. Task will be resubmitted later and continue
+          // the work.
+          return Optional.of(this);
+        }
+
         if (iterator == null) {
           iterator = input.iterator();
         }
 
         while (iterator.hasNext()) {
-          if (queue.size() >= approximateMaxQueueSize) {
```

Comment:
Isn't this a pretty significant change in behavior or ParallelIterable? I assume we have some performance implications of not fully loading the queue?

Reply:
The queue should still be bounded albeit not strictly up to `approximateMaxQueueSize`.

> I assume we have some performance implications of not fully loading the queue?

What do you mean by not fully loading the queue? This PR guarantees that `Task` that is started will complete and not hold external resources.